drivers/xen/gntdev-dmabuf.c

```diff
@@ -88,6 +112,62 @@ struct gntdev_dmabuf_priv {
 
 static void dmabuf_exp_release(struct kref *kref);
 
+static struct gntdev_dmabuf_wait_obj *
+dmabuf_map_wait_obj_find(struct gntdev_dmabuf_priv *priv, int fd)
+{
+	struct gntdev_dmabuf_wait_obj *obj, *ret = ERR_PTR(-ENOENT);
+
+	mutex_lock(&priv->lock);
+	list_for_each_entry(obj, &priv->map_wait_list, next)
+		if (obj->fd == fd) {
+			pr_debug("Found gntdev_dmabuf in the wait list\n");
+			ret = obj;
+			break;
+		}
+	mutex_unlock(&priv->lock);
+
+	return ret;
+}
+
+static void dmabuf_exp_wait_obj_free(struct gntdev_dmabuf_priv *priv,
+			struct gntdev_dmabuf_wait_obj *obj);
+
+static int
+dmabuf_map_wait_obj_set(struct gntdev_dmabuf_priv *priv,
+			struct gntdev_dmabuf *gntdev_dmabuf, int fd)
+{
+	struct gntdev_dmabuf_wait_obj *obj;
+
+	obj = dmabuf_map_wait_obj_find(gntdev_dmabuf->priv, fd);
+	if (IS_ERR(obj) && PTR_ERR(obj) != -ENOENT)
+		dmabuf_exp_wait_obj_free(priv, obj);
+
+	obj = kzalloc(sizeof(*obj), GFP_KERNEL);
+	if (!obj)
+		return -ENOMEM;
+
+	init_completion(&obj->completion);
+	obj->gntdev_dmabuf = gntdev_dmabuf;
+	obj->fd = fd;
+	mutex_lock(&priv->lock);
+	list_add(&obj->next, &priv->map_wait_list);
+	mutex_unlock(&priv->lock);
+	return 0;
+}
+
+static void dmabuf_map_wait_obj_signal(struct gntdev_dmabuf_priv *priv,
+				       struct gntdev_dmabuf *gntdev_dmabuf)
+{
+	struct gntdev_dmabuf_wait_obj *obj;
+
+	list_for_each_entry(obj, &priv->map_wait_list, next)
```

Comment:
Should you take `priv->lock` before traversing the list?

Reply:
I think no need for priv->lock because we don't do any manipulation with list, but made it list_for_each_entry_safe